src/assets/TokensController.ts

```diff
@@ -250,65 +287,170 @@ export class TokensController extends BaseController<
+        const { address } = tokenToIgnore;
+        const checksumAddress = toChecksumHexAddress(address);
+        ignoredTokensMap[address.toLowerCase()] = true;
+        return checksumAddress;
+      });
+      ignoredTokens.push(...formattedTokens);
+      const newDetectedTokens = detectedTokens.filter(
+        (token) => !ignoredTokensMap[token.address.toLowerCase()],
+      );
+      const newTokens = tokens.filter(
+        (token) => !ignoredTokensMap[token.address.toLowerCase()],
+      );
+
+      const { newAllIgnoredTokens, newAllDetectedTokens, newAllTokens } =
+        this._getNewAllTokensState({
+          newIgnoredTokens: ignoredTokens,
+          newDetectedTokens,
+          newTokens,
+        });
+
+      this.update({
+        ignoredTokens,
+        tokens: newTokens,
+        detectedTokens: newDetectedTokens,
+        allIgnoredTokens: newAllIgnoredTokens,
+        allDetectedTokens: newAllDetectedTokens,
+        allTokens: newAllTokens,
+      });
+    } finally {
+      releaseLock();
+    }
+  }
+
+  /**
+   * Adds a batch of detected tokens to the stored token list.
+   *
+   * @param newDetectedTokens - Array of detected tokens to be added or updated.
+   */
+  async addDetectedTokens(newDetectedTokens: Token[]) {
+    const releaseLock = await this.mutex.acquire();
+    const { tokens, detectedTokens, ignoredTokens } = this.state;
+
+    try {
+      newDetectedTokens.forEach((tokenToAdd) => {
+        const { address, symbol, decimals, image, aggregators } = tokenToAdd;
         const checksumAddress = toChecksumHexAddress(address);
         const newEntry: Token = {
           address: checksumAddress,
           symbol,
           decimals,
           image,
-          isERC721,
+          aggregators,
         };
-        const previousEntry = tokens.find(
+        const previousImportedEntry = tokens.find(
           (token) =>
             token.address.toLowerCase() === checksumAddress.toLowerCase(),
         );
-        if (previousEntry) {
-          const previousIndex = tokens.indexOf(previousEntry);
-          tokens[previousIndex] = newEntry;
+        if (previousImportedEntry) {
+          // Update existing data of imported token
+          const previousImportedIndex = tokens.indexOf(previousImportedEntry);
+          tokens[previousImportedIndex] = newEntry;
         } else {
-          tokens.push(newEntry);
+          const ignoredTokenIndex = ignoredTokens.indexOf(address);
+          if (ignoredTokenIndex === -1) {
+            // Add detected token
+            const previousDetectedEntry = detectedTokens.find(
+              (token) =>
+                token.address.toLowerCase() === checksumAddress.toLowerCase(),
+            );
+            if (previousDetectedEntry) {
+              const previousDetectedIndex = detectedTokens.indexOf(
+                previousDetectedEntry,
+              );
+              detectedTokens[previousDetectedIndex] = newEntry;
```

Comment:
This will directly modify state as well. Should `detectedTokens` be copied first?

Reply:
Updated to not directly mutate state. This was happening in a bunch of other places in this controller so I went ahead and fixed those as well.